research/object_detection/core/target_assigner.py

```diff
@@ -1897,3 +1899,170 @@ def assign_corner_offset_targets(
+      cls_targets_list.append(cls_targets)
+      cls_weights_list.append(cls_weights)
+      reg_targets_list.append(reg_targets)
+      reg_weights_list.append(reg_weights)
+    batch_cls_targets = tf.stack(cls_targets_list)
+    batch_cls_weights = tf.stack(cls_weights_list)
+    batch_reg_targets = tf.stack(reg_targets_list)
+    batch_reg_weights = tf.stack(reg_weights_list)
+    return (batch_cls_targets, batch_cls_weights, batch_reg_targets,
+            batch_reg_weights)
+
+  def assign(self,
+             box_preds,
+             groundtruth_boxes,
+             class_predictions,
+             groundtruth_labels,
+             groundtruth_weights=None):
+    """Assign classification and regression targets to each box_pred.
+
+    For a given set of box_preds and groundtruth detections, match box_preds
+    to groundtruth_boxes and assign classification and regression targets to
+    each box_pred as well as weights based on the resulting match (specifying,
+    e.g., which box_preds should not contribute to training loss).
+
+    box_preds that are not matched to anything are given a classification target
+    of self._unmatched_cls_target which can be specified via the constructor.
+
+    Args:
+      box_preds: a BoxList representing N box_preds
+      groundtruth_boxes: a BoxList representing M groundtruth boxes
+      class_predictions: A tensor with shape [max_num_boxes, num_classes]
+        to be used by certain similarity calculators.
+      groundtruth_labels:  a tensor of shape [M, num_classes]
+        with labels for each of the ground_truth boxes. The subshape
+        [num_classes] can be empty (corresponding to scalar inputs).  When set
+        to None, groundtruth_labels assumes a binary problem where all
+        ground_truth boxes get a positive label (of 1).
+      groundtruth_weights: a float tensor of shape [M] indicating the weight to
+        assign to all box_preds match to a particular groundtruth box. The weights
+        must be in [0., 1.]. If None, all weights are set to 1. Generally no
+        groundtruth boxes with zero weight match to any box_preds as matchers are
+        aware of groundtruth weights. Additionally, `cls_weights` and
+        `reg_weights` are calculated using groundtruth weights as an added
+        safety.
+
+    Returns:
+      cls_targets: a float32 tensor with shape [num_box_preds, num_classes],
+        where the subshape [num_classes] is compatible with groundtruth_labels
+        which has shape [num_gt_boxes, num_classes].
+      cls_weights: a float32 tensor with shape [num_box_preds, num_classes],
+        representing weights for each element in cls_targets.
+      reg_targets: a float32 tensor with shape [num_box_preds, box_code_dimension]
+      reg_weights: a float32 tensor with shape [num_box_preds]
+
+    """
+    unmatched_class_label = tf.constant([1] + [0] * (groundtruth_labels.shape[1] - 1), tf.float32)
```

Comment:
ok, sorry, I think I was wrong about this, It should be configurable. the value is different based on whether we use softmax of sigmoid outputs.

Reply:
Why would it be different? Even if using sigmoid, the background class may still be a target, when you want nothing else predicted right?